### PR TITLE
feat: Radionuclides backfill job (#558)

### LINKED ISSUE #558: BDMS-611: Implement Radionuclides backfill job and BDD step definitions
## Summary

The feature spec `tests/features/nma-chemistry-radionuclides-refactor.feature` defines 7 scenarios for refactoring legacy `NMA_Radionuclides` rows into the Ocotillo `Observation` schema via a backfill job.

## Current State

PR #573 (`558-radionuclides-backfill`) implements the full backfill job and BDD step definitions. Remaining work is resolving the local test DB setup (#585) and final review.

### Completed

- [x] Fix test harness — seed `lexicon_term` with required terms in `before_all`
- [x] Implement step definitions — `tests/features/steps/chemistry-backfill.py` covering all 7 scenarios
- [x] Implement backfill job — `transfers/backfill/chemistry_backfill.py`
  - Maps `NMA_Radionuclides` rows → `Observation` records keyed on `nma_pk_chemistryresults` (GlobalID)
  - Creates/reuses `Parameter` records by `(parameter_name, matrix="water")`
  - Sets `detect_flag=false` when `Symbol = "<"`
  - Stores `Volume`/`VolumeUnit` on the parent `Sample` (first-write-wins with conflict warnings)
  - Creates `Notes` records linked to observations
  - Preserves `AnalysisMethod` as-is (no lexicon mapping)
  - Skips orphan records (missing Sample linkage) and reports count
  - Is idempotent (re-run safe)
  - Deterministic row processing via `ORDER BY NMA_Radionuclides.id`
- [x] Register in backfill orchestrator — wired into `transfers/backfill/backfill.py`
- [x] CLI arg rejection — explicit error on removed `--batch-size` flag
- [x] AnalysisMethod test cleanup scoped to test-created IDs (closes #567)
- [x] Chemistry backfill test cleanup handles errors gracefully (closes #566)

### Remaining

- [ ] Resolve test DB setup so `uv run pytest` passes locally (#585)
- [ ] Final code review and merge

## Linked

- PR: #573
- Jira: BDMS-611
- Blockers: #585 (load_dotenv convention)
- Closes: #567, #566
- Feature spec: `tests/features/nma-chemistry-radionuclides-refactor.feature`

### PR DESCRIPTION
**Pull request by kbighorse**
## Summary

- Implements the Radionuclides backfill pipeline that maps `NMA_Radionuclides` → `Observation` records with chemistry metadata
- Adds Alembic migration for new columns on `observation` (`nma_pk_chemistryresults`, `detect_flag`, `uncertainty`, `analysis_agency`) and `sample` (`nma_pk_chemistrysample`, `volume`, `volume_unit`)
- Drops stale `thing_id` FK from `NMA_Radionuclides` (relationships now go through `NMA_Chemistry_SampleInfo`)
- Adds BDD test suite (7 scenarios) covering happy path, idempotency, orphan handling, edge cases
- Adds `pCi/L` unit and `Chemistry Observation` note type to lexicon

### Robustness
- Per-row savepoint isolation — one bad row never aborts the run
- Rows with missing `AnalysisDate` are skipped (logged as errors) instead of using sentinel values
- Migration upgrade is idempotent (checks `information_schema` before dropping columns/FKs)
- Migration downgrade safely restores `thing_id` (nullable → backfill → NOT NULL)
- Full tracebacks logged via `logger.exception()` for per-row failures

### Test safety
- Cleanup scoped to scenario `sample_ids` (no global deletes)
- Analysis method tracking scoped to scenario samples

Closes #558
Refs #565, #566, #567, #568, #570, #571

## Test plan

- [ ] CI passes (BDD scenarios + existing pytest suite)
- [ ] Run `alembic upgrade head` on a fresh DB — migration applies cleanly
- [ ] Run `alembic downgrade -1` — migration rolls back cleanly
- [ ] Run backfill against staging NMA_Radionuclides data
- [ ] Verify idempotency: re-running backfill produces 0 inserts, 0 errors

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/features/steps/chemistry-backfill.py [hunk 1]
```diff
@@ -0,0 +1,703 @@
...
+        samples = (
+            session.execute(select(Sample).where(Sample.volume_unit.isnot(None)))
+            .scalars()
+            .all()
+        )
+        assert len(samples) >= 1, "Expected at least one Sample with volume_unit set"
+        sample = samples[-1]
+        assert (
+            sample.volume_unit == unit
+        ), f"Expected volume_unit='{unit}', got '{sample.volume_unit}'"
```
[the remark this answers is not in the document]
Comment (kbighorse) [reply tests/features/steps/chemistry-backfill.py:570]: Fixed in c1d1b3f1 — both volume and volume_unit assertions now scoped to scenario sample_ids.

File: tests/features/steps/chemistry-backfill.py [hunk 2]
```diff
@@ -0,0 +1,703 @@
...
+        # Find the sample that was just modified
+        samples = (
+            session.execute(select(Sample).where(Sample.volume.isnot(None)))
+            .scalars()
+            .all()
+        )
+        assert len(samples) >= 1, "Expected at least one Sample with volume set"
+        sample = samples[-1]
+        assert (
+            sample.volume == expected
```
[the remark this answers is not in the document]
Comment (kbighorse) [reply tests/features/steps/chemistry-backfill.py:553]: Fixed in c1d1b3f1 — assertions now query by Sample.id.in_(scenario_sample_ids).

File: transfers/backfill/chemistry_backfill.py [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+
+    # Track existing observation keys so we know insert vs update
+    existing_keys = set(
+        row[0]
+        for row in session.execute(
+            select(Observation.nma_pk_chemistryresults).where(
+                Observation.nma_pk_chemistryresults.isnot(None)
+            )
+        ).all()
+    )
```
[the remark this answers is not in the document]
Comment (kbighorse) [reply transfers/backfill/chemistry_backfill.py:136]: Fixed in c1d1b3f1 — existing_keys now lowercases values to match global_id_str normalization.

File: transfers/backfill/chemistry_backfill.py [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+
+    # Query all radionuclides joined with sample info for nma_sample_pt_id
+    legacy_rows = session.execute(
+        select(NMA_Radionuclides, NMA_Chemistry_SampleInfo.nma_sample_pt_id).join(
+            NMA_Chemistry_SampleInfo,
+            NMA_Radionuclides.chemistry_sample_info_id == NMA_Chemistry_SampleInfo.id,
+        )
+    ).all()
+
+    logger.info("Processing %d legacy Radionuclides rows", len(legacy_rows))
```
[the remark this answers is not in the document]
Comment (kbighorse) [reply transfers/backfill/chemistry_backfill.py]: Acknowledged — batched execution with periodic commits is tracked as #570. Current dataset (~2k rows) fits in memory; we'll add yield_per()/batch commits when scaling.

File: transfers/backfill/chemistry_backfill.py [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+        savepoint = session.begin_nested()
+        try:
+            # Get-or-create Parameter
+            param = _get_or_create_parameter(session, analyte, "water")
+
+            # Get-or-create AnalysisMethod
+            analysis_method_id = None
+            if row.analysis_method:
+                am = _get_or_create_analysis_method(session, row.analysis_method)
+                analysis_method_id = am.id
```
[the remark this answers is not in the document]
Comment (kbighorse) [reply transfers/backfill/chemistry_backfill.py:199]: Good suggestion — in-memory caching for Parameter and AnalysisMethod lookups is a natural optimization for the batched execution work in #570.

File: transfers/backfill/chemistry_backfill.py [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+        # Build observation_datetime
+        obs_dt = row.analysis_date
+        if obs_dt is not None:
+            if obs_dt.tzinfo is None:
+                obs_dt = obs_dt.replace(tzinfo=timezone.utc)
+        else:
+            result.errors.append(
+                f"Row GlobalID={global_id_str} has no AnalysisDate — skipping"
+            )
+            continue
```
[the remark this answers is not in the document]
Comment (kbighorse) [reply transfers/backfill/chemistry_backfill.py:187]: Intentional — when AnalysisDate is missing, the row is fundamentally incomplete for creating an Observation (observation_datetime is NOT NULL). Sample volume is populated via the upsert path for rows that do have dates. Rows with volume but no date are tracked as #571.

File: alembic/versions/545a5b77e5e8_add_chemistry_backfill_columns_to_.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+            nullable=True,
+            comment="Agency or lab that performed the analysis",
+        ),
+    )
+    op.create_unique_constraint(
+        "uq_observation_nma_pk_chemistryresults",
+        "observation",
+        ["nma_pk_chemistryresults"],
+    )
+
```
[the remark this answers is not in the document]
Comment (kbighorse) [reply alembic/versions/545a5b77e5e8_add_chemistry_backfill_columns_to_.py:65]: Fair point — the idempotency specifically refers to the thing_id drop (guarded by information_schema check). Column adds and constraints are standard Alembic operations managed by revision tracking. Clarified in PR description.

**Review by jirhiker — COMMENTED** [review]
Looks good from the description. @ksmuczynski should review re database schema

**Comment by ksmuczynski** [comment]
@jirhiker @kbighorse Great, thanks. Will take a closer look as soon as the well inventory ingestion work is completed.

File: tests/features/steps/chemistry-backfill.py [hunk 8]
```diff
@@ -0,0 +1,704 @@
...
+            row[0]
+            for row in session.execute(
+                select(Observation.analysis_method_id).where(
+                    Observation.sample_id.in_(scenario_sample_ids),
+                    Observation.analysis_method_id.isnot(None),
+                )
+            ).all()
+        ]
+        context._backfill_created["analysis_method_ids"] = list(set(obs_am_ids))
+
```
[the remark this answers is not in the document]
Comment (kbighorse) [reply tests/features/steps/chemistry-backfill.py]: Fixed — `_track_created_analysis_methods` now snapshots all AnalysisMethod IDs before the backfill runs and diffs after, so only actually-created methods are tracked for cleanup.

File: tests/features/environment.py [hunk 9]
```diff
@@ -741,23 +746,123 @@ def before_scenario(context, scenario):
...
+                        session.delete(row)
+
+                # Delete Locations created by _ensure_well
+                for lid in created.get("location_ids", []):
+                    row = session.get(Location, lid)
+                    if row:
+                        session.delete(row)
+
+                session.commit()
+        except Exception:
```
[the remark this answers is not in the document]
Comment (kbighorse) [reply tests/features/environment.py:843]: Fixed — `context._backfill_created` is now deleted in a `finally` block after cleanup, preventing accumulation across scenarios.

**Comment by ksmuczynski** [comment]
Stale PR- closing. Will consider reopening when refactoring chemistry data becomes a priority.
